code/headers/hardware_usart.hpp

```diff
@@ -2,76 +2,184 @@
         /// @param b: byte to send
-        void send_byte(const uint8_t &b);
+        void send_byte(const uint8_t &b) {
+            while (!transmit_ready()) {
+            }
+            hardware_usart->US_THR = b;
+        }
 
         /// @brief receive a byte by reading the US_RHR register
         /// @return byte received
-        uint8_t receive_byte();
+        uint8_t receive_byte() {
+            return hardware_usart->US_RHR;
+        }
 
     public:
-        hardware_usart_c(unsigned int baudrate, uart_ports_c usart_port) ;
+        hardware_usart_c(unsigned int baudrate, usart_ports usart_port)
+        : baudrate(baudrate) {
+            if (usart_port == usart_ports::UART_SIZE){
+                HWLIB_PANIC_WITH_LOCATION;
+            }
+
+            const auto &curr = usart[uint8_t(usart_port)];
+
+            hardware_usart = curr.usart;
+            set_peripheral(curr.pio, curr.rx_mask, curr.periph);
+            set_peripheral(curr.pio, curr.tx_mask, curr.periph);
+            PMC->PMC_PCER0 = (0x01 << curr.id);
+
+            disable();
+
+            hardware_usart->US_BRGR = (5241600u / baudrate);
+
+            hardware_usart->US_MR =
+                    UART_MR_PAR_NO | UART_MR_CHMODE_NORMAL | US_MR_CHRL_8_BIT;
+            hardware_usart->US_IDR = 0xFFFFFFFF;
+
+            enable();
+        }
+
         /// @brief char output operator
         ///
         /// Although calling send_byte should do the exact same thing.
         /// In practice useing this fuction is more stable
         /// Especially when the values are repeated.
-        hardware_usart_c &operator<<(uint8_t byte);
+        hardware_usart_c &operator<<(uint8_t byte) {
+            send_byte(byte);
+            return *this;
+        }
 
         /// @brief string output operator
         ///
         /// used for char arrays to quickly send more than one byte
         /// In practice useing this fuction is more stable than repeated use of
         /// send_byte() Especially when the values are repeated.
-        hardware_usart_c &operator<<(const char *c) ;
+        hardware_usart_c &operator<<(const char *c) {
+            for (const char *p = c; *p != '\0'; p++) {
+                send_byte(*p);
+            }
+            return *this;
+        }
 
         /// @brief enables the internal USART controller
-        void enable() override;
+        void enable() override {
+            hardware_usart->US_CR = UART_CR_RXEN | UART_CR_TXEN;
+        }
 
         /// @brief disables the internal USART controller
-        void disable() override;
+        void disable() override {
+            hardware_usart->US_CR =
+                    UART_CR_RSTRX | UART_CR_RSTTX | UART_CR_RXDIS | UART_CR_TXDIS;
+        }
 
         /// @brief send single byte
         /// @param c: byte to send
         /// @return true if byte send, false if not send
-        bool send(const uint8_t c) override;
+        bool send(const uint8_t c) override {
```

Comment:
This never returns false, either the comment and the return type need to be updated or you need to handle the failing of a send.

Reply:
It's also inconsistent with the `putc` function below which doesn't do anything with the return value.

Reply:
will be resolved when issue #20 is addressed